source/util/timer.cpp

```diff
@@ -0,0 +1,105 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <sys/resource.h>
+#include <sys/time.h>
+#include <iomanip>
+#include <iostream>
+
+#include "util/timer.h"
+
+namespace spvutils {
+
+#if defined(SPIRV_TIMER_ENABLED)
+
+namespace {
+inline double TimeDifference(const timeval& before, const timeval& after) {
+  return static_cast<double>(after.tv_sec - before.tv_sec) +
+         static_cast<double>((after.tv_usec - before.tv_usec) / 10000) * .01;
+}
+}  // namespace
+
+void TimerPrintDescription(std::ostream* out) {
+  if (out) {
+    *out << std::setw(30) << "PASS name" << std::setw(12) << "USR time"
+         << std::setw(12) << "WALL time" << std::setw(12) << "SYS time"
+#if defined(SPIRV_MEMORY_MEASUREMENT_ENABLED)
+         << std::setw(12) << "RSS" << std::setw(12) << "Pagefault"
+#endif  // defined(SPIRV_MEMORY_MEASUREMENT_ENABLED)
+         << std::endl;
+  }
+}
+
+void Timer::Start() {
+  if (report_stream_) {
+    if (getrusage(RUSAGE_SELF, &usage_before) == -1) {
+      usage_fail = kGetrusageFail;
+    } else if (gettimeofday(&wall_before, NULL) == -1) {
+      usage_fail = kGettimeofdayFail;
+    }
+  }
+}
+
+void Timer::Stop() {
+  if (report_stream_ && usage_fail == kNotFailed) {
+    if (getrusage(RUSAGE_SELF, &usage_after) == -1) {
+      usage_fail = kGetrusageFail;
+    } else if (gettimeofday(&wall_after, NULL) == -1) {
+      usage_fail = kGettimeofdayFail;
+    }
+  }
+}
+
+void Timer::Report(const char* tag) {
+  switch (usage_fail) {
+    case kGetrusageFail:
+      *report_stream_ << std::setw(30) << tag
+                      << " ERROR: calling getrusage() fails";
+      return;
+    case kGettimeofdayFail:
+      *report_stream_ << std::setw(30) << tag
+                      << " ERROR: calling gettimeofday() fails";
+      return;
+    default:
+      break;
+  }
+
+  if (report_stream_) {
+    report_stream_->precision(2);
+    *report_stream_ << std::fixed << std::setw(30) << tag << std::setw(12)
+                    << TimeDifference(usage_before.ru_utime,
+                                      usage_after.ru_utime)
+                    << std::setw(12) << TimeDifference(wall_before, wall_after)
+                    << std::setw(12)
+                    << TimeDifference(usage_before.ru_stime,
+                                      usage_after.ru_stime)
+#if defined(SPIRV_MEMORY_MEASUREMENT_ENABLED)
+                    << std::setw(12)
+                    << (usage_after.ru_maxrss - usage_before.ru_maxrss)
+                    << std::setw(12)
+                    << ((usage_after.ru_minflt - usage_before.ru_minflt) +
+                        (usage_after.ru_majflt - usage_before.ru_majflt))
+#endif  // defined(SPIRV_MEMORY_MEASUREMENT_ENABLED)
+                    << std::endl;
+  }
+}
+
+void Timer::StopAndReport(const char* tag) {
```

Comment:
Yeah, this method does not give us enough extra convenience.  I would remove it.

Reply:
Done